pkg/services/datasources/service/datasource_service.go

```diff
@@ -218,7 +214,18 @@ func (s *Service) UpdateDataSource(ctx context.Context, cmd *models.UpdateDataSo
 		}
 	}
 
-	return s.SecretsStore.Set(ctx, cmd.OrgId, cmd.Name, secretType, string(secret))
+	if len(cmd.SecureJsonData) > 0 {
```

Comment:
I think you need to mimic the old behavior, see https://github.com/grafana/grafana/blob/0af09cd0815bfe16e9ea9ce3c2e1ce40775a4455/pkg/api/datasources.go#L342-L367

The old behavior basically merged updated secure JSON data keys with the existing stored secure JSON data.

Reply:
> Since the keys are not merged when `len(cmd.SecureJsonData) == 0` I think we still need a check for `len(cmd.SecureJsonData) != 0` before updating the secrets. 

Don't really see any difference between `if len(cmd.SecureJsonData) > 0 {` and `if len(cmd.SecureJsonData) != 0 {` in datasource_service.go so don't understand that change - it can't be negative length, right?

> I did update `fillWithSecureJSONData` to use the `DecryptedValues` function on [d921f59](https://github.com/grafana/grafana/commit/d921f59e52056134f40f49a6fb033ae1bafe3f69).
> Do you think it would be best to keep the check for `!= 0` or remove the one that doesn't merge the keys when `len(..) == 0` ?

It's probably because it historically took for granted secureJsonData will not be updated in the database if it's empty/nil - An XORM thing. Not sure what's best, maybe safer to skip the `if len(cmd.SecureJsonData) == 0` check and always merge.

When looking at it now, maybe the fillSecureJsonData logic should be located in the datasource_service.go instead to support modifying datasources from multiple places rather than living in the API layer, but not sure. Maybe an improvement for later. WDYT?

Sorry if I provided confusing answers but was a bit unsure what code you referred to in some cases :smile: You did reference the actual code/commit above and that is great :+1:, removes any confusion.

Reply:
It does makes sense to have that code on datasource_service.go instead.

Reply:
Done on f800dd2

Reply:
Looks good :+1: